sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriPropertiesShouldNotBeStrings.cs

```diff
@@ -0,0 +1,72 @@
+/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class UriPropertiesShouldBeStrings : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3996";
+        private const string MessageFormat = "Change this property type to 'System.Uri'";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected sealed override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var propertyDeclaration = (PropertyDeclarationSyntax)c.Node;
+                    var propertySymbol = c.SemanticModel.GetDeclaredSymbol(propertyDeclaration);
+
+                    if (propertySymbol.Type.SpecialType != SpecialType.System_String)
+                    {
+                        return;
+                    }
+
+                    if (!NameContains(propertySymbol.Name, "uri", "urn", "url"))
```

Comment:
Not C# fluent enough to be 100% sure of the following : but I would rather have an external set to call contains on it. To avoid creation of array on every single call of this lambda.
And I would tend to prefer positive condition in this case : `if(IsCheckedName(propertySymbol.Name)) {` the lambda should not know about the checked words.